esiclient/v1/orchestrator/cluster.py

```diff
@@ -0,0 +1,224 @@
+            print("* Using port %s" % port_name)
+        return port
+
+    def provision_node(self, node, provisioning_type, node_config):
+        glance_client = self.app.client_manager.image
+        ironic_client = self.app.client_manager.baremetal
+        neutron_client = self.app.client_manager.network
+
+        network_config = node_config['network']
+
+        # create network port
+        port = self.get_port_from_network_config(node, network_config)
+
+        # provision
+        if provisioning_type == 'image':
+            image = glance_client.find_image(
+                node_config['provisioning']['image_uuid'])
+            ssh_key = node_config['provisioning'].get('ssh_key', None)
+            if ssh_key is None:
+                raise ESIOrchestrationException(
+                    "ssh_key must be specified for image provisioning")
+            print("* Provisioning node %s with image %s" % (
+                node.name, image.name))
+            utils.provision_node_with_image(
+                node.uuid, node.resource_class, port.id, image.id,
+                ssh_key)
+        elif provisioning_type == 'image_url':
+            url = node_config['provisioning'].get('url', None)
+            if url is None:
+                raise ESIOrchestrationException(
+                    "url must be specified for image URL provisioning")
+            print("* Provisioning node %s from url %s" % (
+                node.name, url))
+            utils.boot_node_from_url(node.uuid, url, port.id, ironic_client)
+
+        if 'fip_network_uuid' in network_config:
+            print("* Assigning floating IP to node %s on port %s" % (
+                node.name, port.name))
+            fip_network = neutron_client.find_network(
+                network_config['fip_network_uuid'])
+            utils.get_or_assign_port_floating_ip(
+                port, fip_network, neutron_client)
+
+        return node, port
+
+    def take_action(self, parsed_args):
+        self.log.debug("take_action(%s)", parsed_args)
+
+        cluster_config_file = parsed_args.cluster_config_file
+        with open(cluster_config_file) as f:
+            cluster_config = json.load(f)
+
+        self.assign_nodes(cluster_config)
+
+        print("")
+
+        print("PROVISIONING NODES")
+        node_configs = cluster_config['node_configs']
+        futures = []
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            for node_config in node_configs:
+                provisioning_type = \
+                    node_config['provisioning']['provisioning_type']
+                if provisioning_type not in self.PROVISIONING_METHODS:
+                    raise ESIOrchestrationException(
+                        "Unknown provisioning method %s" % provisioning_type)
+                nodes = node_config['nodes']['ironic_nodes']
+                for node in nodes:
+                    future = executor.submit(
```

Comment:
I wonder do we need a try-expection block here? Also cleanup resources (delete the port or set the node to available if there's an error happens)

Reply:
Can you print more information during node provisioning? Such as which node is being deployed. It might be useful when the execution time is too long and user has no idea about what's the current stage.

Reply:
Oh, there are messages printed within the `provision_node` function.

I'm a bit undecided upon the other questions. Right now, if multiple nodes are provisioned and only one node has an error, the others will still provision to completion. I'm not sure I actually want to delete resources in that case. I think I'm also against auto-cleaning because it takes away the ability of the user to investigate the issue.

In any case, exceptions for individual node issues end up printed in stdout, which I - think? - is fine. I catch ESI-specific exceptions, and let the others just bubble up.

Reply:
Yeah, I agree.